### DISCUSSION TITLE
Generic Connection trait

### DISCUSSION THREAD

**Discussion by MarcAntoine-Arnaud** [diesel-rs/diesel#2803] [Q&A]
Hello,

I'm starting to use Diesel, thanks for this great project !
I'm just a little bit disappointed regarding Connection. Currently everything need to be implement fro `PgConnection`, `SqliteConnection`...
I expected it can be a generic trait that allow a generic usage of queries.

Do you think it can be realist/complex ? Or just not needed right now ?
Or already available but not very well documented or I have not look at the right place ?

In my case I wanna to develop one code that support by default Sqlite, but with a configuration switch Postgres can also be used for the same application.

Thank you,
Marc-Antoine

**Comment by weiznich** [comment]
There is already a [`Connection`](https://docs.diesel.rs/1.4.x/diesel/connection/trait.Connection.html) trait in diesel. In combination with the `RunQueryDsl` and `QueryDsl` this can be used to write code that is generic over more tgan one connection type. See diesel-migrations and  diesel-cli for examples.

That written, there is not much documentation for this topic as we normally discourage users going down this way as this leads to:
* really complex where clauses
* loosing the ability to use backend specific sql extensions to improve perfomance


Improvements in this area are welcome, but will require quite substantial contributions.
